core/tests/protractor_utils/waitFor.js

```diff
@@ -181,6 +181,33 @@ var fileToBeDownloaded = async function(filename) {
   }, DEFAULT_WAIT_TIME_MSECS, 'File was not downloaded!');
 };
 
+var clientSideRedirection = async function(action, check) {
+  // Client side redirection is known to cause "both angularJS testability
+  // and angular testability are undefined" flake.
+  // As suggested by protractor devs here (http://git.io/v4gXM), waiting for
+  // angular is disabled during client side redirects.
+  await browser.waitForAngularEnabled(false);
+
+  // Action triggering redirection.
+  await action();
+
+  // The action only triggers the redirection but does not wait for it to
+  // complete. Manually waiting for redirection here.
+  await browser.driver.wait(async() => {
+    var url = await browser.driver.getCurrentUrl();
+    // Condition to wait on.
+    return check(url);
+  }, DEFAULT_WAIT_TIME_MSECS);
+
+  // The above solution only waits until url is changed, so invoking
+  // pageToFullyLoad to wait for all async tasks on new page to finish before
+  // proceeding.
+  await pageToFullyLoad();
```

Comment:
> wait for all async tasks on new page to finish

This isn't what `pageToFullyLoad()` does. It only waits for full-page loading screens to disappear. You can see its implementation here:

https://github.com/oppia/oppia/blob/bd43b9dbd3e1680ff9e14223a1ad2f8e0000a827/core/tests/protractor_utils/waitFor.js#L63-L72

It looks like you're trying to wait for the new page to finish loading before exiting this function, but there's no generic way to wait for all async tasks to finish. (side note: protractor normally handles this, but its automatic waiting doesn't work with our hybrid state of both Angular and AngularJS) I think you need the caller of the function to specify what to wait for

Reply:
I think your updated comment is still not quite right:

>  The above solution only waits until url is changed, so invoking pageToFullyLoad to wait for the new page to load compeletly.

Invoking `pageToFullyLoad` waits for the full-page loading screen to disappear. This does not guarantee that the page has loaded completely for a couple reasons:

1. Not all pages use a full-page loading screen. Some use smaller loading indicators, for example ellipses in the navbar.
2. Even on pages that use a full-page loading screen, the page sometimes hasn't fully loaded even after the screen disappears. For example, the toolbars of rich-text editors sometimes appear after the full-page loading screen disappears.

Basically, `pageToFullyLoad` is an unfortunately-named function that doesn't really wait for the page to be fully loaded.

Reply:
Yes, `waitForAngularEnabled(true)` will be called just after the URL changes.
So, there can be a case that browser is still loading javascript bundles and angular is not initialized on the page during the call to waitForAngularEnabled(true). So, we cannot remove waiting.

Made the changes according to first approach.